### PR TITLE
Port existing E4K changes from previews/E4Kadapter branch

### PR DESCRIPTION
**Pull request by timtay-microsoft**
- Send connect/disconnect packet after open/before close
- Add flag for E4K mode
- Create Mqtt5 layer when in E4K mode, otherwise create the existing Mqtt3 layer

### CODE REVIEW

File: iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt5/Mqtt5IotHubConnection.java [hunk 1]
```diff
@@ -206,7 +227,45 @@ else if (proxySettings.getProxy().type() == Proxy.Type.HTTP)
...
+            {
+                log.debug("Sending the \"Opened a connection\" message to the MQTT broker");
+                E4KConnectionMessagePayload connectPayload = E4KConnectionMessagePayload.builder()
+                    .connectionState(E4KConnectionState.Connected)
+                    .modelId(this.config.getModelId())
+                    .mqttVersion(MQTT_VERSION_STRING)
+                    .deviceClientType(this.config.getProductInfo().getUserAgentString())
+                    .build();
+                byte[] serializedPayload = GSON.toJson(connectPayload).getBytes(StandardCharsets.UTF_8);
+                org.eclipse.paho.client.mqttv3.IMqttDeliveryToken token = this.mqttAsyncClient.publish(
```
Comment (brycewang-microsoft) [inline iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt5/Mqtt5IotHubConnection.java:240]: Does Mqtt5 have an object type similar to `mqttv3.IMqttDeliveryToken` here?

Comment (timtay-microsoft) [reply iothub/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/mqtt5/Mqtt5IotHubConnection.java]: Good catch. This was a copy paste error. I have pushed the fix and now this is an Mqtt5 object

File: iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/E4KConnectionMessagePayloadTest.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+public class E4KConnectionMessagePayloadTest
+{
+    @Test
+    public void TestSerialization()
+    {
+        String expectedDeviceClientType = "some device client type";
+        String expectedMqttVersion = "3.1.1";
```
Comment (brycewang-microsoft) [inline iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/E4KConnectionMessagePayloadTest.java:16]: I am wondering if we should tie this to v5 only now that E4K requires Mqtt 5?

Comment (timtay-microsoft) [reply iothub/device/iot-device-client/src/test/java/com/microsoft/azure/sdk/iot/device/transport/mqtt/E4KConnectionMessagePayloadTest.java:16]: Yes

**Review by brycewang-microsoft — APPROVED** [review]

**Pull request merged**